### ISSUE TITLE
Datalog Instantiation issues

### ISSUE THREAD

**Issue by lanfangping** [ravel-net/pyotr#7]
## Datatype mismatch when instantiating a rule in Faure-log

The issue occurs when running the route aggregation test case in `unit_tests.py`. When instantiating the rule `R(1,x)[x == 10] :- l(1,2), l(1,3), l(1,4), R(2,x)[x == 10]`, two attributes in R are int4_faure datatype, the insert SQL is `insert into R values(2,'x','{"x == 10"}')`. The datatype mismatch happens at `2`, i.e., we should add single quotes to `2`. The error shows as follows.
```
Traceback (most recent call last):
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/unit_tests.py", line 98, in <module>
    if (not program2.contains(program1)):
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/program.py", line 65, in contains
    if not self.contains_rule(rule):
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/program.py", line 108, in contains_rule
    rule2.addConstants(conn)
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/rule.py", line 191, in addConstants
    atom.addConstants(conn, self._mapping)
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/atom.py", line 192, in addConstants
    cursor.execute(sql)
psycopg2.errors.DatatypeMismatch: column "c0" is of type int4_faure but expression is of type integer
LINE 1: insert into R values(2,'x','{"x == 10"}')
                             ^
HINT:  You will need to rewrite or cast the expression.
```
## An invalid array representation in insert SQL in Datalog
Consider a rule `R(a3,h1,[a3, z, y, x],4) :- R(c1,h1,[z, y, x],3),l(a3,c2)`. When treat it as a data instance, the generated insertion is `insert into R values(0,1,'{2, 3, 4}',3)`. `'{2, 3, 4}'` is an invalid representation in the insert SQL. The error shows as follows.

```
Checking  R(a3,h1,[a3, z, y, x],4) :- R(c1,h1,[z, y, x],3),l(a3,c2)
Traceback (most recent call last):
  File "/home/fangping/Projects/pyotr/experiments/minimization_datalog/fattree/scalability_datalog/script.py", line 43, in <module>
    avg_time, total_atoms, total_rules = test_scalability(r, k, 'h1')
  File "/home/fangping/Projects/pyotr/experiments/minimization_datalog/fattree/scalability_datalog/script.py", line 25, in test_scalability
    program.minimize()
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/program.py", line 148, in minimize
    contained = self.contains_rule(rule_with_deleted_atom)
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/program.py", line 108, in contains_rule
    rule2.addConstants(conn)
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/rule.py", line 191, in addConstants
    atom.addConstants(conn, self._mapping)
  File "/home/fangping/Projects/pyotr/Core/Homomorphism/Datalog/atom.py", line 192, in addConstants
    cursor.execute(sql)
psycopg2.errors.InvalidTextRepresentation: invalid input syntax for type integer: "{2, 3, 4}"
LINE 1: insert into R values(0,1,'{2, 3, 4}',3)
                                 ^
```

**Comment by mudbri** [comment]
@lanfangping, have you tried recompiling int4_faure in the latest commit? This should be fixed

**Comment by mudbri** [comment]
Was fixed in 078de2ec21c0b019cad1730556f6656604ab210b. Closing issue
